File: src/main/java/com/capstone/backend/service/ProjectService.java
package com.capstone.backend.service;

import com.capstone.backend.dto.ProjectRequestDto;
import com.capstone.backend.dto.ProjectResponseDto;
import com.capstone.backend.entity.Project;
import com.capstone.backend.entity.ProjectTechStack;
import com.capstone.backend.entity.TechStack;
import com.capstone.backend.entity.User;
import com.capstone.backend.repository.ProjectRepository;
import com.capstone.backend.repository.ProjectTechStackRepository;
import com.capstone.backend.repository.TechStackRepository;
import com.capstone.backend.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Set;
import com.capstone.backend.dto.MemberResponseDto;
import com.capstone.backend.entity.Application;
import com.capstone.backend.repository.ApplicationRepository;
import com.capstone.backend.global.exception.ForbiddenException;

@Service
@RequiredArgsConstructor
public class ProjectService {

    private static final Set<String> ALLOWED_STATUSES = Set.of("모집중", "진행중", "완료됨");

    private final UserRepository userRepository;
    private final ProjectRepository projectRepository;
    private final TechStackRepository techStackRepository;
    private final ProjectTechStackRepository projectTechStackRepository;
    private final ApplicationRepository applicationRepository;

    @Transactional
    public ProjectResponseDto createProject(ProjectRequestDto request, String email) {
        User leader = userRepository.findByEmailIgnoreCase(email)
                .orElseThrow(() -> new IllegalArgumentException("유저를 찾을 수 없습니다."));

        Project project = Project.builder()
            .title(request.getTitle())
            .content(request.getContent())
            .region(request.getRegion())
            .status("모집중")  // ← 이것만 남기고
            .term(String.valueOf(request.getTerm()))
            .leader(leader)
            .meetingType(request.getMeetingType())
            .isLocalOnly(request.getIsLocalOnly())
            // .status(request.getStatus())  ← 이 줄 삭제!
            .build();

        Project savedProject = projectRepository.save(project);

        if (request.getTechStackIds() != null && !request.getTechStackIds().isEmpty()) {
            List<TechStack> techStacks = techStackRepository.findAllById(request.getTechStackIds());

            List<ProjectTechStack> projectTechStacks = techStacks.stream()
                    .map(techStack -> ProjectTechStack.builder()
                            .project(savedProject)
                            .techStack(techStack)
                            .build())
                    .toList();

            projectTechStackRepository.saveAll(projectTechStacks);
        }

        return ProjectResponseDto.from(savedProject);
    }

    // 1. 전체 프로젝트 조회 (페이징 버전으로 변경)
    public Page<ProjectResponseDto> getAllProjects(Pageable pageable) {
        return projectRepository.findAll(pageable)
                .map(ProjectResponseDto::from);
    }

    public ProjectResponseDto getProjectById(Long projectId) {
    return projectRepository.findById(projectId)
            .map(ProjectResponseDto::from)
            .orElseThrow(() -> new IllegalArgumentException("프로젝트를 찾을 수 없습니다."));
    }

    @Transactional
    public ProjectResponseDto updateProject(Long projectId, ProjectRequestDto request, String email) {
        Project project = projectRepository.findById(projectId)
            .orElseThrow(() -> new IllegalArgumentException("프로젝트를 찾을 수 없습니다."));
        requireLeader(project, email);

        project.updateTitle(request.getTitle());
        project.updateContent(request.getContent());
        project.updateRegion(request.getRegion());
        project.updateTerm(String.valueOf(request.getTerm()));
        project.updateMeetingType(request.getMeetingType());
        project.updateIsLocalOnly(request.getIsLocalOnly());

        return ProjectResponseDto.from(project);
    }

    @Transactional
    public void deleteProject(Long projectId, String email) {
        Project project = projectRepository.findById(projectId)
                .orElseThrow(() -> new IllegalArgumentException("프로젝트를 찾을 수 없습니다."));
        requireLeader(project, email);
        projectRepository.delete(project);
    }

    public List<MemberResponseDto> getProjectMembers(Long projectId) {
        return applicationRepository.findByProjectIdAndStatus(projectId, "ACCEPTED")
                .stream()
                .map(app -> new MemberResponseDto(
                        app.getApplicant().getId(),
                        app.getApplicant().getName(),
                        app.getSupportRole()
                ))
                .toList();
    }

    @Transactional
    public ProjectResponseDto updateProjectStatus(Long projectId, String status, String email) {
        Project project = projectRepository.findById(projectId)
            .orElseThrow(() -> new IllegalArgumentException("프로젝트를 찾을 수 없습니다."));
        requireLeader(project, email);
        if (!ALLOWED_STATUSES.contains(status)) {
            throw new IllegalArgumentException("프로젝트 상태가 올바르지 않습니다.");
        }

    project.updateStatus(status);

    return ProjectResponseDto.from(project);
    }

    @Transactional
    public void removeProjectMember(Long projectId, Long memberId, String email) {
    Project project = projectRepository.findById(projectId)
            .orElseThrow(() -> new IllegalArgumentException("프로젝트를 찾을 수 없습니다."));
    requireLeader(project, email);
    Application application = applicationRepository
        .findByProjectIdAndApplicantIdAndStatus(projectId, memberId, "ACCEPTED")
        .orElseThrow(() -> new IllegalArgumentException("해당 팀원을 찾을 수 없습니다."));
    applicationRepository.delete(application);
    }

    @Transactional
    public void addProjectMember(Long projectId, Long userId, String email) {
    Project project = projectRepository.findById(projectId)
            .orElseThrow(() -> new IllegalArgumentException("프로젝트를 찾을 수 없습니다."));
    requireLeader(project, email);

    User user = userRepository.findById(userId)
            .orElseThrow(() -> new IllegalArgumentException("유저를 찾을 수 없습니다."));

    Application application = Application.builder()
            .project(project)
            .applicant(user)
            .supportRole("팀장")
            .message("")
            .status("ACCEPTED")
            .build();

    applicationRepository.save(application);
    }

    private void requireLeader(Project project, String email) {
        if (!project.getLeader().getEmail().equalsIgnoreCase(email)) {
            throw new ForbiddenException("프로젝트 팀장만 수행할 수 있습니다.");
        }
    }

}

File: frontend/src/services/projectService.js
import api from "./api";
const API_URL = "/api/projects";
const APPLICATION_API = "/api/application";

export const projectService = {
    // 💡 [수정 포인트] MainPage에서 넘겨주는 page와 size 파라미터를 받아 axios 쿼리 스트링으로 전달합니다.
    getAllProjects: async (page = 0, size = 9) => {
        try {
            // Axios의 params 옵션을 사용하면 자동으로 URL 뒤에 ?page=0&size=9 를 붙여줍니다.
            const res = await api.get(API_URL, {
                params: {
                    page: page,
                    size: size
                }
            });
            return res.data;
        } catch (e) {
            console.error(e);
            // 에러 발생 시 MainPage의 크래시를 방지하기 위해 페이징 객체 기본 구조를 반환합니다.
            return { content: [], totalPages: 1 };
        }
    },
    getProjectById: async (projectId) => {
        try {
            const res = await api.get(`${API_URL}/${projectId}`);
            return res.data;
        } catch (e) {
            console.error(e);
            return null;
        }
    },
    createProject: async (data) => {
        const res = await api.post(API_URL, {
            title: data.title,
            content: data.content,
            region: data.region,
            status: "모집중",
            term: data.term,
            techStackIds: data.techStackIds,
            meetingType: data.meetingType,
            isLocalOnly: data.isLocalOnly
        });
        return res.data;
    },
    updateProject: async (projectId, data) => {
        const res = await api.put(`${API_URL}/${projectId}`, {
            title: data.title,
            content: data.content,
            region: data.region,
            status: data.status,
            term: data.term,
            techStackIds: data.techStackIds,
            meetingType: data.meetingType,
            isLocalOnly: data.isLocalOnly
        });
        return res.data;
    },
    deleteProject: async (projectId) => {
        try {
            await api.delete(`${API_URL}/${projectId}`);
            return true;
        } catch (e) {
            console.error(e);
            return false;
        }
    },
    applyToProject: async (projectId) => {
        await api.post(APPLICATION_API, {
            project_id: projectId,
            support_role: "지원자",
            message: "",
            status: "PENDING"
        });
        return true;
    },
    cancelApplication: async (applicationId) => {
        await api.delete(`${APPLICATION_API}/${applicationId}`);
        return true;
    },
    getProjectMembers: async (projectId) => {
        const res = await api.get(`${API_URL}/${projectId}/members`);
        return Array.isArray(res.data) ? res.data : [];
    },
    updateProjectStatus: async (projectId, status) => {
        const res = await api.patch(`${API_URL}/${projectId}/status`, { status });
        return res.data;
    },
    removeProjectMember: async (projectId, memberId) => {
        const res = await api.delete(`${API_URL}/${projectId}/members/${memberId}`);
        return res.data;
    },
    addProjectMember: async (projectId, userId) => {
        const res = await api.post(`${API_URL}/${projectId}/members`, {
            user_id: userId
        });
        return res.data;
    },
};
